cep-checker/main.go

```diff
@@ -0,0 +1,249 @@
+	pods, err := getPods(ctx, clientset)
+	if err != nil {
+		return nil, err
+	}
+
+	ceps, err := getCeps(ctx, dynamicClient)
+	if err != nil {
+		return nil, err
+	}
+
+	missings := make(map[string]string)
+
+	// pod -> cep
+	for key := range pods {
+		if _, ok := ceps[key]; !ok {
+			// To avoid a miss detection, check again
+			res, err := check(ctx, clientset, dynamicClient, key)
+			if err != nil {
+				return nil, err
+			}
+			if res != "" {
+				missings[key] = res
+			}
+		}
+	}
+
+	// cep -> pod
+	for key := range ceps {
+		if _, ok := pods[key]; !ok {
+			// To avoid a miss detection, check again
+			res, err := check(ctx, clientset, dynamicClient, key)
+			if err != nil {
+				return nil, err
+			}
+			if res != "" {
+				missings[key] = res
+			}
+		}
+	}
+	return missings, nil
+}
+
+// Check checks the consistency for given key(namespace/name).
+// If there is the inconsistency(which one is missing), output a log and create a metric.
+func check(ctx context.Context, clientset *kubernetes.Clientset, dynamicClient *dynamic.DynamicClient, key string) (string, error) {
+
+	podExist := false
+	cepExist := false
+
+	s := strings.Split(key, "/")
+	ns := s[0]
+	name := s[1]
+
+	_, err := clientset.CoreV1().Pods(ns).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return "", err
+		}
+	} else {
+		podExist = true
+	}
+
+	_, err = dynamicClient.Resource(schema.GroupVersionResource{Group: CILIUM_GROUP, Version: CILIUM_VERSION, Resource: CEP_NAME}).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return "", err
+		}
+	} else {
+		cepExist = true
+	}
+
+	if podExist != cepExist {
+		if podExist {
+			log.WarnContext(ctx, "find a missing CEP", slog.String("namespace", ns), slog.String("name", name))
+			m := fmt.Sprintf(`%s{name="%s", namespace="%s", resource="cep"}`, METRICS_MISSING_NAME, name, ns)
+			metrics.GetOrCreateGauge(m, func() float64 { return 1 })
+			return "cep", nil
+		}
+		if cepExist {
+			log.WarnContext(ctx, "find a missing Pod", slog.String("namespace", ns), slog.String("name", name))
+			m := fmt.Sprintf(`%s{name="%s", namespace="%s", resource="pod"}`, METRICS_MISSING_NAME, name, ns)
+			metrics.GetOrCreateGauge(m, func() float64 { return 1 })
+			return "pod", nil
+		}
+	}
+
+	return "", nil
+}
+
+// list pods that is not in host network and that is running all namespaces
+func getPods(ctx context.Context, clientset *kubernetes.Clientset) (map[string]struct{}, error) {
+	pods := make(map[string]struct{})
+
+	podList, err := clientset.CoreV1().Pods("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pod := range podList.Items {
+		if pod.Spec.HostNetwork {
+			continue
+		}
+		if pod.Status.Phase == "Succeeded" {
+			continue
+		}
+		if pod.Status.Phase == "Pending" && pod.Status.PodIP == "" {
+			continue
+		}
```

Comment:
When a pod's status is "Failed" or "Unknown", it passes this section.
I think it is enough to check if a pod is "Running" and has an IP address.
WDYT?

Reply:
When pods are in "Succeeded" phase, CEP is already deleted as far as I checked.
So I think we need to check this phase, but it may be enough for other cases to check pods are in "Running" and have IP addresses.

Reply:
How about checking like this?

```go
		if pod.Status.Phase != "Running" || pod.Status.PodIP != "" {
			continue
		}
```